controller/lib/debug/trace.h

```diff
@@ -22,35 +22,61 @@ limitations under the License.
 
 #define TRACE_VAR_CT 4
 
-// Called from the main loop
-void TraceSample();
-
-// local functions
-int CountActiveVars(DebugVar *vptr[]);
-
 // Trace buffer
 extern CircBuff<uint32_t, 0x4000> traceBuffer;
 
-// The trace control variable is a bit-mapped parameter which
-// is used to start/stop the trace and read its status.
-// Right now only bit 0 is in use.  Set this to start capturing
-// data to the trace buffer.  It will auto-clear when the buffer
-// is full
-extern int32_t traceCtrl;
+class TraceControl {
+public:
+  uint32_t GetFlags() const { return flags_; }
+  void SetFlags(uint32_t flags) {
+    if ((flags & 1) && !(flags_ & 1)) {
+      traceBuffer.Flush();
+    }
+    flags_ = flags;
+  }
+
+  uint32_t GetPeriod() { return period_; }
+  void SetPeriod(uint32_t period) { period_ = period; }
+
+  void MaybeSample() {
+    if (!(flags_ & 1))
+      return;
+    if (++count_ < period_)
+      return;
+    count_ = 0;
+
+    if (!SampleAllVars()) {
+      // Trace buffer is full, stop tracing.
+      SetFlags(0);
+    }
+  }
 
-// The trace period gives the period of the trace data capture
-// in units of loop cycles.
-extern int32_t tracePeriod;
+  void Flush() {
+    flags_ = 0;
+    while (traceBuffer.Get() != std::nullopt) {
+    }
+  }
```

Comment:
Seems like an awfully inefficient way to flush the buffer.  That could loop 16 thousand times.  Why not just call traceBuffer.Flush()?

Reply:
Yeah, this was copied from [here](https://github.com/RespiraWorks/VentilatorSoftware/blob/1875bcc63403ff2a111d3de8f803f42ca84594b3/controller/lib/debug/trace_cmd.cpp#L44). Fixed.